examples/pytorch/dimenet/qm9.py

```diff
@@ -0,0 +1,218 @@
+    
+    def download(self):
+        """ step 2 """
+        file_path = f'{self.raw_dir}/qm9_eV.npz'
+        if not os.path.exists(file_path):
+            download(self._url, path=file_path)
+
+    def process(self):
+        """ step 3 """
+        npz_path = f'{self.raw_dir}/qm9_eV.npz'
+        data_dict = np.load(npz_path, allow_pickle=True)
+        # data_dict['N'] contains the number of atoms in each molecule.
+        # Atomic properties (Z and R) of all molecules are concatenated as single tensors,
+        # so you need this value to select the correct atoms for each molecule.
+        self.N = data_dict['N']
+        self.R = data_dict['R']
+        self.Z = data_dict['Z']
+        self.N_cumsum = np.concatenate([[0], np.cumsum(self.N)])
+        # graph labels
+        self.label_dict = {}
+        for k in self._keys:
+            self.label_dict[k] = F.tensor(data_dict[k], dtype=F.data_type_dict['float32'])
+
+        self.label = F.stack([self.label_dict[key] for key in self.label_keys], dim=1)
+        # graph features
+        self.graphs, self.line_graphs = self._load_graph()
+    
+    def _load_graph(self):
+        num_graphs = self.label.shape[0]
+        graphs = []
+        line_graphs = []
+        
+        for idx in trange(num_graphs):
+            n_atoms = self.N[idx]
+            R = self.R[self.N_cumsum[idx]:self.N_cumsum[idx + 1]]
+            dist = np.linalg.norm(R[:, None, :] - R[None, :, :], axis=-1)
+            adj = sp.csr_matrix(dist <= self.cutoff) - sp.eye(n_atoms, dtype=np.bool)
+            adj = adj.tocoo()
+            u, v = F.tensor(adj.row), F.tensor(adj.col)
+            g = dgl_graph((u, v))
+            g.ndata['R'] = F.tensor(R, dtype=F.data_type_dict['float32'])
+            g.ndata['Z'] = F.tensor(self.Z[self.N_cumsum[idx]:self.N_cumsum[idx + 1]], 
+                                    dtype=F.data_type_dict['int64'])
+            
+            # add user defined features
+            if self.edge_funcs is not None:
+                for func in self.edge_funcs:
+                    g.apply_edges(func)
+
+            graphs.append(g)
+            l_g = dgl.line_graph(g, backtracking=False)
+            line_graphs.append(l_g)
+    
+        return graphs, line_graphs
+
+    def save(self):
+        """ step 4 """
+        graph_path = f'{self.save_path}/dgl_graph.bin'
+        line_graph_path = f'{self.save_path}/dgl_line_graph.bin'
+        save_graphs(str(graph_path), self.graphs, self.label_dict)
+        save_graphs(str(line_graph_path), self.line_graphs)
+
+    def load(self):
+        """ step 5 """
+        graph_path = f'{self.save_path}/dgl_graph.bin'
+        line_graph_path = f'{self.save_path}/dgl_line_graph.bin'
+        self.graphs, label_dict = load_graphs(graph_path)
+        self.line_graphs, _ = load_graphs(line_graph_path)
+        self.label = F.stack([label_dict[key] for key in self.label_keys], dim=1)
```

Comment:
Do we need `label_dict` and `self.label`? You seem to be loading labels from the raw file anyway.

Reply:
`self.label` is a subset of `label_dict`, while `label_dict` is from a local file.